### DISCUSSION TITLE
Consider moving to modern Java (17 LTS)

### DISCUSSION THREAD

**Discussion by KvanTTT** [antlr/antlr4#3369] [General]
Java 7 support reaches the end in the middle of 2022: https://en.wikipedia.org/wiki/Java_version_history. Moreover, it's extended support, not even the base.

I suggest moving to the latest modern LTS Java, namely Java 17 that was released in September 2021. Also, it contains a lot of new features and API that could greatly improve developers' performance.

**Comment by ericvergnaud** [comment]
Most users want to get bug fixes and improvements without making big jumps such as changing jvm.
I support dropping Java 7 in favor of Java 8 only.

**Reply by KvanTTT** [reply]
Anyway, Java 8 is also much better than Java 7, but also is outdated.

**Comment by ericvergnaud** [comment]
There would also be a benefit in releasing a Java 9 version in order fo benefit from Java modules.

**Reply by KvanTTT** [reply]
Java 9 is a bad choice because it's not LTS and it's already not supported. The better version is Java 11.

**Comment by parrt** [comment]
At this point I'm ready to move to a newer version of java... Getting this to compile is driving me crazy!  Perhaps we tell people that need much earlier versions to use 4.9.3 or earlier?

**Reply by KvanTTT** [reply]
Yes, just write it to release notes.

**Comment by parrt** [comment]
Ok, so I've been playing around and I'm using java 8, the lowest version I could find on Oracle's website. Everything works fine using source and target 1.7 so we might as well leave it. I remember the android people or something needed an old version of Java so why upgrade if we don't have to. We can build with any version we want but the target can be left at 1.7 I think.

**Reply by ericvergnaud** [reply]
Keeping Java 7 stops us from modernizing the code. So I would recommend dropping it.

**Reply by KvanTTT** [reply]
It will be possible to use one of the latest versions (say 17 or at least 11) for building but with any target >= 1.7? Do I understand correctly?

**Reply by KvanTTT** [reply]
Otherwise, Java 7 syntax is too inconvenient and verbose, it's quite tediously to write on it. It looks like talking to a simpleton. I also recommend dropping it.

**Comment by parrt** [comment]
Which is min version you recommend? jdk 8?  that's easy to do as i have https://github.com/antlr/antlr4/pull/3392 already in progress.

**Reply by parrt** [reply]
Ah. Eric said previously "I support dropping Java 7 in favor of Java 8 only."  @KvanTTT let's go with 8 and see how that goes. We can always go higher next time if we feel the need.

**Reply by parrt** [reply]
by the way I will leave the target as 7 but we can set the source minimum as 8.

**Reply by parrt** [reply]
shit. looks like 1.8 must gen 1.7 at least in maven. :(   Maybe we go to java 11?  I'll try it.

**Reply by parrt** [reply]
hahah. i'll keep trying

```
[INFO] Compiling 218 source files to /Users/parrt/antlr/code/antlr4/runtime-testsuite/target/test-classes
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by org.antlr.v4.test.runtime.CommentHasStringValueProcessor (file:/Users/parrt/antlr/code/antlr4/runtime-testsuite/processors/target/antlr4-runtime-test-annotation-processors-4.9.4-SNAPSHOT.jar) to field com.sun.tools.javac.model.JavacElements.javaCompiler
```

**Reply by parrt** [reply]
java 11 seems a no go. can't use javac internals?

```
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project antlr4-runtime-test-annotation-processors: Compilation failure: Compilation failure: 
[ERROR] /Users/parrt/antlr/code/antlr4/runtime-testsuite/processors/src/org/antlr/v4/test/runtime/CommentHasStringValueProcessor.java:[9,31] error: package com.sun.tools.javac.main does not exist
[ERROR] /Users/parrt/antlr/code/antlr4/runtime-testsuite/processors/src/org/antlr/v4/test/runtime/CommentHasStringValueProcessor.java:[10,32] error: package com.sun.tools.javac.model does not exist
[ERROR] /Users/parrt/antlr/code/antlr4/runtime-testsuite/processors/src/org/antlr/v4/test/runtime/CommentHasStringValueProcessor.java:[11,31] error: package com.sun.tools.javac.tree does not exist
[ERROR] /Users/parrt/antlr/code/antlr4/runtime-testsuite/processors/src/org/antlr/v4/test/runtime/CommentHasStringValueProcessor.java:[12,31] error: package com.sun.tools.javac.tree does not exist
```

this is after adding:

```
                        <compilerArgs>
                            <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.main=ALL-UNNAMED</arg>
                            <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.util=ALL-UNNAMED</arg>
                            <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.model=ALL-UNNAMED</arg>
                            <arg>-J--add-exports=jdk.compiler/com.sun.tools.javac.tree=ALL-UNNAMED</arg>
                        </compilerArgs>
```

**Reply by parrt** [reply]
ha!! This guy built a tools.jar dependency:

```
<dependency>
  <groupId>com.github.olivergondza</groupId>
  <artifactId>maven-jdk-tools-wrapper</artifactId>
  <version>0.1</version>
</dependency>
```

**Comment by parrt** [comment]
Pushed new change to mac/appveyor...hoping it'll fix. off to teach for 2 hours.

**Reply by KvanTTT** [reply]
It looks like [fixed on AppVeyor](https://ci.appveyor.com/project/parrt/antlr4/builds/41806809/job/o0f9bbp872g4t1l1), but there are the following errors during building:

```
[ERROR] 
Maven dependencies of Maven Plugins should be in provided scope.
Please make sure that all your dependencies declared in POM whose group ID is
org.apache.maven have set '<scope>provided</scope>' as well.
In the future this error will break the build.
The following dependencies are in wrong scope:
 * org.apache.maven:maven-plugin-api:jar:3.8.4:compile
 * org.apache.maven:maven-model:jar:3.8.4:compile
 * org.apache.maven:maven-artifact:jar:3.8.4:compile
 * org.apache.maven:maven-settings:jar:3.8.4:compile
 * org.apache.maven:maven-repository-metadata:jar:3.8.4:compile
 * org.apache.maven:maven-project:jar:2.2.1:compile
 * org.apache.maven:maven-profile:jar:2.2.1:compile
 * org.apache.maven:maven-artifact-manager:jar:2.2.1:compile
 * org.apache.maven:maven-plugin-registry:jar:2.2.1:compile
Please fix your build!
[ERROR]
```

But it's still [failing on Mac](https://github.com/antlr/antlr4/runs/4449260365?check_suite_focus=true):

```
Error:  Failed to execute goal io.takari.maven.plugins:takari-lifecycle-plugin:2.0.7:testProperties (testProperties) on project antlr4-maven-plugin: The plugin io.takari.maven.plugins:takari-lifecycle-plugin:2.0.7 requires Maven version [3.6.3,) -> [Help 1]
Error:  
Error:  To see the full stack trace of the errors, re-run Maven with the -e switch.
Error:  Re-run Maven using the -X switch to enable full debug logging.
Error:  
Error:  For more information about the errors and possible solutions, please read the following articles:
Error:  [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/PluginIncompatibleException
Error:  
Error:  After correcting the problems, you can resume the build with the command
Error:    mvn <goals> -rf :antlr4-maven-plugin
Error: Process completed with exit code 1.
```

**Comment by parrt** [comment]
Yeah, weird. that exact thing works on my mac. Will mess with `plugins:takari-lifecycle-plugin` versions I guess.

**Reply by parrt** [reply]
I can't fix the `Please fix your build!` as I don't know maven but just pushed fix I hope for takari bullshit.

**Reply by parrt** [reply]
Works much better but still screwed up in mac build at github:

```
Error:  Failed to execute goal org.antlr:antlr4-maven-plugin:4.9.4-SNAPSHOT:antlr4 (default) on project antlr4-runtime-testsuite: The plugin org.antlr:antlr4-maven-plugin:4.9.4-SNAPSHOT requires Maven version 3.8 -> [Help 1]
```

I don't know how to fix this. (And it works locally on mac.).  This is a circular dependency.  Ideas @KvanTTT ?

**Reply by parrt** [reply]
Oh duh. github uses `3.5.4` so i'll change setup.

**Comment by parrt** [comment]
Ok, we're back to passing! Well, linux times out on parsing tests for C++. :(   Should we remove "Travis-CI Build Status (Swift-Linux)" since it seems out of date and we test swift on mac?

**Comment by parrt** [comment]
Ok, I have been able to get this to build now with 1.8 and 11 but with source and target 1.7 in the maven pom, so I'm going to leave it. See https://github.com/antlr/antlr4/pull/3392   lots of incompatibility between the different versions and I don't want to make a maven file that is really sensitive to the Java version.  This project is super complicated with all of the use of the internal javac tools.jar stuff that went away and now this module crap in java...ugh.

**Reply by KvanTTT** [reply]
Maybe it makes sense to move on to something else, not outdated in modern versions of Java.

**Reply by parrt** [reply]
Would be great to be able to build in java 17, even if we leave the source and target low.    The problem is there appears to be a compatibility issue. If we move to something modern, then we force the use of java 17 to build/test.  also, I don't want to do the work myself to replace all of the javac tools.jar code.

**Reply by KvanTTT** [reply]
This compatibility issue is related to annotations in runtimes tests, right?

**Reply by parrt** [reply]
yep. We use the annotations to process comments that contain the actual tests and then we generate the tests for each target.

**Reply by KvanTTT** [reply]
Is it possible to use file-based data instead of comment-based? Also, it should be more convenient because of correct syntax highlighting.

**Reply by parrt** [reply]
I'm sure it's possible... I can't remember now what our reasoning was but here's what they look like now:

```
	public static class ComplementSet extends BaseParserTestDescriptor {
		public String input = "a";
		public String output = null;
		/**
		line 1:0 token recognition error at: 'a'
		line 1:1 missing {} at '<EOF>'
		 */
		@CommentHasStringValue
		public String errors;

		public String startRule = "parse";
		public String grammarName = "T";

		/**
		 grammar T;
		 parse : ~NEW_LINE;
		 NEW_LINE: '\\r'? '\\n';
		 */
		@CommentHasStringValue
		public String grammar;
	}
```

So, it would mean making maybe a json file or something instead of a Java class with test methods...

**Reply by KvanTTT** [reply]
> I can't remember now what our reasoning was but here's what they look like now:

Probably it was not clear how to implement it in the best way using a single file for everything.

> So, it would mean making maybe a json file or something instead of a Java class with test methods...

JSON is not very convenient. I suggest just using a plain file with different parts (grammar, errors, input, output, etc.) separated by special comments:

```
//$ Input: a

//$ startRule: parse

//$ grammarName: T

//$ errors
line 1:0 token recognition error at: 'a'
line 1:1 missing {} at '<EOF>'

//$ grammar
grammar T;
parse : ~NEW_LINE;
NEW_LINE: '\\r'? '\\n';
```

We use a similar structure for Kotlin compiler tests.

**Reply by parrt** [reply]
something like that could work although it would mean a massive rewrite of the test rig... seems like it would be a lot of work for a project that is in maintenance mode.

**Reply by KvanTTT** [reply]
But I think such conversion can be automated.

**Reply by parrt** [reply]
oh that is true!! it gens tests now but could be modified to regenerate itself in text files. (I agree that xml/json and friends are such overkill.)

**Reply by KvanTTT** [reply]
I can try next year if you don't mind such changes.

**Reply by parrt** [reply]
First, I'd like to hear what @ericvergnaud thinks but it ain't a hard no ;)

**Reply by KvanTTT** [reply]
And it's possible to use ANTLR-generated parser to parse Java files.

**Comment by ericvergnaud** [comment]
Originally the tests were in separate files, but that created a bit of a complexity coz you needed 4 files: grammar, input, errors, output
The comments approach make it possible to guarantee that a change only affects the corresponding test, which I like.
It also makes it encouraging to write more tests.

I’m pretty sure we can get this working with Java 8 for the runtime and Java 17 for the tool.
Right now I’m a bit busy, but happy to join efforts with Ivan sometime in the future.

**Reply by parrt** [reply]
Ok, let's put off for now.

**Reply by KvanTTT** [reply]
BTW, Java 16 introduced [text blocks](https://docs.oracle.com/en/java/javase/16/text-blocks/index.html) that allow convenient multiline strings. Maybe it's even more preferable to use them. If we can target Java 8 but use the recent Java version for development I suggest using them especially if we are failed with internal API.

**Reply by parrt** [reply]
that does look ideal and is in fact what I think I originally wanted (text blocks) but the only way to fake that was with some complicated mechanism that looked at the comments.

I guess that would be really minimal intrusion into the source base.  We could use Java 8 to process the `BaseRuntimeTestDescriptor` class definitions so that instead of:

```
/**
line 1:0 token recognition error at: 'a'
line 1:1 missing {} at '<EOF>'
*/
@CommentHasStringValue
public String errors;
```

they would be:

```
public String errors = """
line 1:0 token recognition error at: 'a'
line 1:1 missing {} at '<EOF>'
"""
```

I wonder if the javac stuff inside tools.jar can alter the source code and write it back out, as it is processing the annotations. Do you have any idea @KvanTTT ?   If that works, then it should be a fairly easy conversion process and then we would be able to use a modern version of Java for building.

**Reply by KvanTTT** [reply]
I don't have an idea yet. Actually, I learned about text blocks today from the article: https://reflectoring.io/java-release-notes/

**Reply by parrt** [reply]
haha. cool.  Heh, do you guys know https://bazel.build/ ?   The Google guys are offering to donate build files... i'd love to use a build system that my small brain understands. if this Bazel is any good...

**Reply by KvanTTT** [reply]
I haven't heard about it. But I'm not sure it's able to easily resolve all our problems with building.

**Reply by parrt** [reply]
Agreed.  a separate issue.

**Reply by parrt** [reply]
I jumped over to fix my tensor-sensor lib today but I can think about this testing thing again soon. also I have to grade exams starting tomorrow.

**Reply by KvanTTT** [reply]
I work on weekdays, but I'm mostly free on weekends. Firstly, I would like to enable C++ test on Windows (https://github.com/antlr/antlr4/issues/3389) and refactor runtime test infrastructure (https://github.com/antlr/antlr4/issues/3367).

**Comment by parrt** [comment]
Ok, my new galactic overlords can only use Java 11 at this time and must build from src not use jar. I'll rebuild tests to use resource files of some flavor and set src=11, target=8.

**Comment by jcking** [comment]
IMO the preferred stance would be allowing ANTLR to require the previous Java LTS version. Today the most recent LTS is Java 17 with the previous LTS being Java 11, thus ANTLR should be free to require Java 11. This creates a happy medium.

**Reply by parrt** [reply]
Yep, i'm rebuilding test rig now to remove dependence on javac tools.jar and will set src to 11, target to 8.

**Comment by parrt** [comment]
Ok, merged 3rd gen test rig. Compiles with java 17 it appears but set at 1.8 src/trg. Happy to have someone figure out how to make this 11 src, 7 or 8 trg.

**Comment by parrt** [comment]
Just to keep this thread up to date: current build after I merged the test rig update we are stuck at version 8 java for source and target.

**Reply by Techcable** [reply]
When this is done do you plan to include a `module-info.java` file for the runtime? I wanna use `jlink` for my app but that's difficult to do without modules :(

Also, It seems that recent java versions don't play well with maven on my Mac 😦 I can't even get antlr to build.

**Reply by KvanTTT** [reply]
We moved only to Java 8. As I understand modules are supported since 9. Does your info relate to it?

> Also, It seems that recent java versions don't play well with maven on my Mac 😦 I can't even get antlr to build.

Is Java 8 supported?

**Reply by parrt** [reply]
@Techcable What specifically is going wrong with the maven build?

**Reply by Techcable** [reply]
> @Techcable What specifically is going wrong with the maven build?

Upon further investigation, it seems that a good deal of my problems seem to be related to not having the other language SDKs installed (dart, C#, python2, etc). Strangely, it seems like a lot of the tests fail too (even for the Java runtime).

Updating antlr (`git pull`) & maven seems to have fixed most of the  isues with the build itself 🤷 Some tests fail but it doesn't clearly seem related to the build system 😄 

If you're planning to move to Java 11, does that mean you are planning to add a `module-info.java`?

**Reply by ericvergnaud** [reply]
We've moved the tool to Java 11 but kept the runtime as Java 8 to avoid breaking things (Java 8 is till widely used), so no module-info

**Comment by parrt** [comment]
@Techcable Yeah, I only test a few targets locally since I don't have all of those languages installed... I rely on the continuous integration to handle those other tests.
